### PR TITLE
[clang-cl] Fix value of __FUNCTION__ in MSVC mode.

### PR DESCRIPTION
**Pull request by zahiraam**
Predefined macro FUNCTION in clang is not returning the same string than MS for templated functions.

See https://godbolt.org/z/q3EKn5zq4

For the same test case MSVC is returning:

function: TestClass::TestClass
function: TestStruct::TestStruct
function: TestEnum::TestEnum

The initial work for this was in the reverted patch (https://github.com/llvm/llvm-project/pull/66120). This patch solves the issues raised in the reverted patch.

### CODE REVIEW

File: clang/lib/AST/TypePrinter.cpp
Comment (Sirraide) [inline clang/lib/AST/TypePrinter.cpp:2277]:
```suggestion
        OS << Argument.getAsType().getAsString();
```

**Review by Sirraide — CHANGES_REQUESTED** [review]

**Comment by Sirraide** [comment]
I’m not entirely sure this change is a good idea: some people might be relying on Clang’s behaviour wrt `__FUNCTION__`, especially with how long it’s been around; Clang may not be fully MSVC-compatible in this case, but that incompatibilty seems unlikely to cause problems imo since afaik most people that actually depend on the output of these sorts of macros for e.g. reflexion use `__FUNCSIG__` or `__PRETTY_FUNCTION__` instead—then again, that could also be a reason as to why changing it might actually be fine.

**Comment by zahiraam** [comment]
> I’m not entirely sure this change is a good idea: some people might be relying on Clang’s behaviour wrt `__FUNCTION__`, especially with how long it’s been around; Clang may not be fully MSVC-compatible in this case, but that incompatibilty seems unlikely to cause problems imo since afaik most people that actually depend on the output of these sorts of macros for e.g. reflexion use `__FUNCSIG__` or `__PRETTY_FUNCTION__` instead—then again, that could also be a reason as to why changing it might actually be fine.

Thanks for the review. This is an issue that we encountered in our internal testing with an application using `__FUNCTION__`. I agree that might not be used by a lot of users/apps, but it is used and having it compatible with MSVC doesn't actually hurt? (I hope).

File: clang/test/Analysis/eval-predefined-exprs.cpp
Comment (Sirraide) [reply clang/test/Analysis/eval-predefined-exprs.cpp:63]: Reading this again, it seems it’s because of the relative `@`. I’d suggest using markers instead, or if possible, putting the comments on the same line as the diagnostic to avoid problems like these—I know this is how it’s done in the rest of the file too, but we tend to prefer not to use relative offsets in cases like these.

**Comment by Sirraide** [comment]
Oh, I just noticed that you’ve tried to land this change before (https://github.com/llvm/llvm-project/pull/66120), but it seems it was reverted because of some problems; if those are fixed now, then I’d say going forward with this is probably fine, seeing as it was approved the first time round.

This would have been a bit less confusing if you had pointed that out, just saying ;Þ

**Comment by zahiraam** [comment]
> Oh, I just noticed that you’ve tried to land this change before (#66120), but it seems it was reverted because of some problems; if those are fixed now, then I’d say going forward with this is probably fine, seeing as it was approved the first time round.
> 
> This would have been a bit less confusing if you had pointed that out, just saying ;Þ

Yes. Sorry about that. This patch was mostly to fix the issues raised in PR#66120.

File: clang/test/SemaCXX/source_location.cpp
Comment (zahiraam) [reply clang/test/SemaCXX/source_location.cpp:485]: Yes CL output is: `TestClass<class UnitTestNative>::TestClass`

File: clang/test/Analysis/eval-predefined-exprs.cpp
Comment (Sirraide) [inline clang/test/Analysis/eval-predefined-exprs.cpp:59]:
```suggestion
    clang_analyzer_dump(__func__);     // expected-warning {{&Element{"A",0 S64b,char}}}
```
The `#ifdef` and `#else` branch seem identical to me; if that’s not a copy-and-paste error, then I don’t think we need the same check twice.

File: clang/test/Analysis/eval-predefined-exprs.cpp
Comment (Sirraide) [inline clang/test/Analysis/eval-predefined-exprs.cpp:83]:
```suggestion
    clang_analyzer_dump(__func__);          // expected-warning {{&Element{"~A",0 S64b,char}}}
```
As above

File: clang/docs/ReleaseNotes.rst
Comment (Sirraide) [inline clang/docs/ReleaseNotes.rst:228]:
```suggestion
- Fix value of predefined macro ``__FUNCTION__`` in MSVC compatibility mode.
  Fixes (#GH66114).
```
I seem to have missed this earlier, but we should probably point out that this change only applies to `MSVCCompat` mode (and also use the new issue link syntax that we have now).

**Review by Sirraide — CHANGES_REQUESTED** [review]

**Review by Sirraide — APPROVED** [review]

**Comment by Sirraide** [comment]
Alright, the changes look sensible to me, but I’d like for someone who’s more familiar with this part of Clang or with the original pr to comment on them too before we merge it.

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:228]:
```suggestion
  Fixes (#GH66114).
```

File: clang/lib/AST/Expr.cpp
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:732]:
```suggestion
    bool IsFuncOrFunctionInNonMSVCCompatEnv =
        ((IK == PredefinedIdentKind::Func ||
          IK == PredefinedIdentKind ::Function) &&
         !LO.MSVCCompat);
    bool IsLFunctionInMSVCCommpatEnv =
        IK == PredefinedIdentKind::LFunction && LO.MSVCCompat;
    bool IsFuncOrFunctionOrLFunctionOrFuncDName =
```
Coding style nits, NFC

File: clang/lib/AST/Expr.cpp
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:739]:
```suggestion
        (!ForceElaboratedPrinting && isFuncOrFunctionOrLFunctionOrFuncDName))
```
You may need this to avoid diagnostics about mixed `&&` and `||` operators.

File: clang/lib/AST/TypePrinter.cpp
Comment (AaronBallman) [inline clang/lib/AST/TypePrinter.cpp:1646]: Don't we want to reset these once we're done with the call to `printBefore()`?

Comment (zahiraam) [reply clang/lib/AST/TypePrinter.cpp:1648]: Yes, I think so.

Comment (AaronBallman) [reply clang/lib/AST/TypePrinter.cpp:1648]: Hmmm, your changes don't look quite right to me. The `return printBefore(...);` is going to skip right over the restoration.

I think you wanted something more like:
```
      bool OldTag = Policy.SuppressTagKeyword, OldScope = Policy.SuppressScope;
      Policy.SuppressTagKeyword = true;
      Policy.SuppressScope = false;
      printBefore(T->getNamedType(), OS);
      Policy.SuppressTagKeyword = OldTag;
      Policy.SuppressScope = OldScope;
      return;
```
Before making the changes, I think you should also add test coverage to the unit test where these changes will fix the test.

Comment (zahiraam) [reply clang/lib/AST/TypePrinter.cpp:1648]: Not sure what you mean here "where these changes will fix the test"? will fix which test? The unit test you are referring to is `DeclPrinterTest`? this should be `TestCXXRecordDecl12`.

Comment (zahiraam) [reply clang/lib/AST/TypePrinter.cpp:1648]: Hmmm! Can't seem to find a test case that fails. At each step of  the printing  `ElaboratedTypePolicyRAII PolicyRAII(Policy);` gets called and that re-initializes the values of` Policy.SuppressTagKeyword` and  `Policy.SuppressScope` to true!

**Comment by Sirraide** [comment]
From what I can tell, it seems that Aaron is a lot more familiar w/ this part of Clang than me, so I’ll leave it to him to comment on whether these changes make sense.

File: clang/lib/AST/DeclPrinter.cpp
Comment (AaronBallman) [inline clang/lib/AST/DeclPrinter.cpp:873]: This check is duplicated in `AddPrefix()`, same below (probably can remove the check from `AddPrefix()` or turn it into an assertion).

File: clang/lib/AST/DeclPrinter.cpp
Comment (AaronBallman) [inline clang/lib/AST/DeclPrinter.cpp:878]: This is only necessary because of the changes you made in TypePrinter.cpp, right? e.g., if we reset the policy back to its original values there, we don't need to do it here.

File: clang/lib/AST/DeclPrinter.cpp
Comment (AaronBallman) [inline clang/lib/AST/DeclPrinter.cpp:1055]: Same here.

File: clang/lib/AST/TypePrinter.cpp
Comment (AaronBallman) [reply clang/lib/AST/TypePrinter.cpp:1641]: You need to reset the values here, or use a separate `PrintingPolicy` object so that you're not changing the value for the default printing policy used.

File: clang/lib/AST/DeclPrinter.cpp
Comment (AaronBallman) [inline clang/lib/AST/DeclPrinter.cpp:875]:
```suggestion
      if (!Policy.SuppressTagKeyword && Policy.SuppressScope &&
          !Policy.SuppressUnwrittenScope)
        MaybePrintTagKeywordIfSupressingScopes(Policy, AFT->getReturnType(),
                                               Out);
      AFT->getReturnType().print(Out, Policy, Proto);
```

File: clang/lib/AST/DeclPrinter.cpp
Comment (AaronBallman) [inline clang/lib/AST/DeclPrinter.cpp:1048]:
```suggestion
  if (!Policy.SuppressTagKeyword && Policy.SuppressScope &&
      !Policy.SuppressUnwrittenScope)
    MaybePrintTagKeywordIfSupressingScopes(Policy, T, Out);
  printDeclType(T, Name);
```

**Review by AaronBallman — COMMENTED** [review]
I spotted some simplifications that can be applied. If precommit CI comes back green, then I think this is basically ready to go.

**Comment by zahiraam** [comment]
Thank you all for the reviews.

**Pull request merged**
